### PR TITLE
bnsd/username: use gconf to manage username validation rules

### LINKED ISSUE #952: Update namespace validation rules
The current username implementation is using a regexp to ensure only certain letters are allowed.

https://github.com/iov-one/weave/blob/master/cmd/bnsd/x/username/username.go#L67-L70

Change those rules to be fulfill new requirements described below.

## Acceptance criteria
- [ ] A username consists of a name and namespace separated by a single `*` character (for example `alice*iov`)
- [ ] A namespace must be `iov`
- [ ] A name must be between 0 and not longer than 64 characters
- [ ] A name must consist of only the following ASCII characters 
  - letters `abcdefghijklmnopqrstuvwxyz`
  - digits  `0123456789`
  - special characters `-._`

### PR DESCRIPTION
**Pull request by husio**
Instead of using hardcoded set of validation rules for the username, use
gconf package to manage validation rules dynamically.
Using gconf has another advantage that we can change the validation
rules during the lifetime of a network without a downtime. Because
changing the validation rule is part of the blockchain, no additional
code is required to support this scenario.

resolve #952

### CODE REVIEW

**Comment by iov-bot-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please include a CHANGELOG entry. <br />
You can find it at <a href="https://github.com/iov-one/weave/blob/master/CHANGELOG.md">CHANGELOG.md</a>.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: cmd/bnsd/x/username/configuration.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+	if err := validateRegexp(c.ValidUsernameLabel); err != nil {
+		errs = errors.AppendField(errs, "ValidUsernameLabel", err)
+	}
+	return nil
+}
+
+// validateRegexp returns an error if provided string is not a valid regular
+// expression.
+// This function ensures that the regular expression is a complete match test
+// by ensuring ^ and $ presence.
```
Comment (webmaster128) [inline cmd/bnsd/x/username/configuration.go]: 👍

File: cmd/bnsd/x/username/handlers_test.go [hunk 2]
```diff
@@ -83,6 +84,14 @@ func TestRegisterTokenHandler(t *testing.T) {
 			db := store.MemStore()
 			migration.MustInitPkg(db, "username")
 
+			config := Configuration{
+				ValidUsernameName:  `[a-z0-9\-_.]{3,64}`,
```
Comment (webmaster128) [inline cmd/bnsd/x/username/handlers_test.go]: escape necessary?

File: cmd/bnsd/x/username/handlers_test.go [hunk 3]
```diff
@@ -281,6 +360,14 @@ func TestChangeTokenTargetHandler(t *testing.T) {
...
```
Comment (webmaster128) [inline cmd/bnsd/x/username/handlers_test.go]: escape necessary?

File: cmd/bnsd/x/username/init_test.go [hunk 4]
```diff
@@ -14,6 +14,13 @@ import (
 func TestGenesisInitializer(t *testing.T) {
 	const genesis = `
 	{
+		"conf": {
+			"username": {
+				"valid_username_name": "^[a-z0-9\\-_.]{3,64}$",
```
Comment (webmaster128) [inline cmd/bnsd/x/username/init_test.go]: escape necessary?

File: cmd/bnscli/testdata/config/genesis.json [hunk 5]
```diff
@@ -12,45 +12,92 @@
...
+      "cash": {
+        "collector_address": "seq:dist/revenue/1",
+        "minimal_fee": "0.1 IOV"
+      },
+      "migration": {
+        "admin": "E28AE9A6EB94FC88B73EB7CBD6B87BF93EB9BEF0"
+      },
+      "username": {
+        "owner": "cond:admin/admin/0000000000000001",
+        "valid_username_name": "^[a-z0-9\\-_.]{3,64}$",
```
[the remark this answers is not in the document]
Comment (husio) [reply cmd/bnscli/testdata/config/genesis.json]: Yes, this is required.

File: cmd/bnsd/app/testdata/fixtures/app.go [hunk 6]
```diff
@@ -79,12 +82,17 @@ func appStateGenesis(keyAddress weave.Address) []byte {
...
+				MinimalFee:       coin.NewCoin(0, 10000, "FRNK"),
 			},
-			"migration": dict{
-				"admin": "seq:multisig/usage/1",
+			"migration": migration.Configuration{
+				Admin: mustParseAddr("seq:multisig/usage/1"),
+			},
+			"username": username.Configuration{
+				Owner:              mustParseAddr("seq:uname/admin/1"),
+				ValidUsernameName:  `^[a-z0-9\-_.]{3,64}$`,
```
[the remark this answers is not in the document]
Comment (husio) [reply cmd/bnsd/app/testdata/fixtures/app.go]: Yes.

File: cmd/bnsd/client/main_test.go [hunk 7]
```diff
@@ -96,6 +97,10 @@ func initGenesis(filename string, addr weave.Address) error {
 			"migration": migration.Configuration{
 				Admin: weave.Condition("multisig/usage/0000000000000001").Address(),
 			},
+			"username": username.Configuration{
+				ValidUsernameName:  `^[a-z0-9\-_.]{3,64}`,
```
[the remark this answers is not in the document]
Comment (husio) [reply cmd/bnsd/client/main_test.go]: yes

File: cmd/bnsd/client/testdata/genesis.json [hunk 8]
```diff
@@ -1,128 +1,134 @@
...
- }
+  "app_hash": "",
+  "app_state": {
+    "conf": {
+      "migration": {
+        "admin": "E28AE9A6EB94FC88B73EB7CBD6B87BF93EB9BEF0"
+      },
+      "username": {
+        "owner": "cond:admin/admin/0000000000000001",
+        "valid_username_name": "^[a-z0-9\\-_.]{3,64}$",
```
[the remark this answers is not in the document]
Comment (husio) [reply cmd/bnsd/client/testdata/genesis.json]: yes

File: cmd/bnsd/scenarios/bnsdtest/bnsdtest.go [hunk 9]
```diff
@@ -247,13 +248,18 @@ func initGenesis(t testing.TB, env *EnvConf, filename string) {
...
-				"minimal_fee":       env.AntiSpamFee,
+			"cash": cash.Configuration{
+				CollectorAddress: mustParseAddr(t, "seq:dist/revenue/1"),
+				MinimalFee:       env.AntiSpamFee,
 			},
 			"migration": dict{
 				"admin": "seq:multisig/usage/1",
 			},
+			"username": username.Configuration{
+				ValidUsernameName:  `^[a-z0-9\-_.]{3,64}$`,
```
[the remark this answers is not in the document]
Comment (husio) [reply cmd/bnsd/scenarios/bnsdtest/bnsdtest.go]: yes

File: cmd/bnsd/x/username/handlers_test.go (hunk 2 above)
Comment (husio) [reply cmd/bnsd/x/username/handlers_test.go]: yes

File: cmd/bnsd/x/username/handlers_test.go (hunk 3 above)
Comment (husio) [reply cmd/bnsd/x/username/handlers_test.go]: yes

File: cmd/bnsd/x/username/init_test.go (hunk 4 above)
Comment (husio) [reply cmd/bnsd/x/username/init_test.go]: yes

**Comment by webmaster128** [comment]
> what about providing the ability to validate the regex when updating it through the proposal?

I would see this kind of testing as part of the voting process (off chain)

**Comment by ruseinov** [comment]
> I would see this kind of testing as part of the voting process (off chain)

Fair, but then we need to provide a testing cli tool that uses exactly the same code as our validation rules in gconf, otherwise it might not yield the same results. 
If we don't want to address it in this PR, which is fair, we need to create a ticket and discuss this to have a proper way forward, otherwise this is likely to be a pain point.

File: cmd/bnsd/x/username/configuration.go [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+	if len(rx) > 1024 {
+		return errors.Wrap(errors.ErrInput, "too long")
+	}
+	_, err := regexp.Compile(rx)
+	if err != nil {
+		return errors.Wrap(errors.ErrInput, err.Error())
+	}
+
+	if rx[0] != '^' || rx[len(rx)-1] != '$' {
+		return errors.Wrap(errors.ErrInput, "regular expression must match the whole input, start with ^ and end with $ characters to enforce full match")
```
Comment (ruseinov) [inline cmd/bnsd/x/username/configuration.go]: nit: `a full match`

**Comment by husio** [comment]
@ruseinov raised a valid point with the regular expression validation. This version does not allow to break the system, but it is possible to create a proposal for a message that will always be rejected. This scenario is not unique to username extension and can happen with any message. For example, I can create a proposal to send -10 IOV from my account.

**Pull request merged**